### PR TITLE
feat(resharding): implemented error handling and test for it

### PR DESCRIPTION
**Pull request by wacban**
- added a new ReshardingStatus - Failed - and use it when resharding fails
- made a number of delays and timeouts configurable so that it's testable
- added a new nayduck test for checking the error handling
- added a bunch of new debug, warn and error logs

https://nayduck.near.org/#/run/3270

### CODE REVIEW

File: chain/chain/src/resharding.rs [hunk 1]
```diff
@@ -233,26 +233,47 @@ impl Chain {
...
         }
-        tries.get_state_snapshot(prev_prev_hash).is_err_and(|err| match err {
-            SnapshotError::SnapshotNotFound(_) => true,
-            SnapshotError::LockWouldBlock => true,
-            SnapshotError::IncorrectSnapshotRequested(_, _) => false,
-            SnapshotError::Other(_) => false,
-        })
+
+        let state_snapshot = tries.get_state_snapshot(prev_prev_hash);
+        if let Err(err) = state_snapshot {
```
Comment (shreyan-gupta) [inline chain/chain/src/resharding.rs:247]: Was the old code too convoluted to understand?

File: chain/client/src/sync/state.rs [hunk 2]
```diff
@@ -1041,7 +1041,7 @@ impl StateSync {
             shard_id,
             state_split_scheduler,
         )?;
-        tracing::debug!(target: "sync", %shard_id, %sync_hash, ?me, "State sync split scheduled");
+        tracing::debug!(target: "sync", %shard_id, %sync_hash, ?me, "resharding scheduled");
```
Comment (shreyan-gupta) [inline chain/client/src/sync/state.rs:1044]: aah, how did we miss this in the past?

File: tools/database/src/corrupt.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#[derive(Parser)]
+pub(crate) struct CorruptStateSnapshotCommand {}
+
+impl CorruptStateSnapshotCommand {
+    pub(crate) fn run(&self, home: &PathBuf) -> anyhow::Result<()> {
+        let store = open_state_snapshot(home, near_store::Mode::ReadWrite)?;
+        let flat_storage_manager = FlatStorageManager::new(store.clone());
+
+        let mut store_update = store.store_update();
+        // TODO(resharding) there must be a better way to get the shard uids
```
Comment (shreyan-gupta) [inline tools/database/src/corrupt.rs:16]: This highly depends on what shard version we are trying to corrupt right? Could we pass in the shard_version as input here maybe and use that to get the shard_uids?

**Review by shreyan-gupta — APPROVED** [review]

File: chain/chain/src/resharding.rs (hunk 1 above)
Comment (wacban) [reply chain/chain/src/resharding.rs:247]: not really, I just wanted to add that debug line in the middle of it

File: tools/database/src/corrupt.rs (hunk 3 above)
Comment (wacban) [reply tools/database/src/corrupt.rs:16]: Not a bad workaround at all, though I prefer to keep the cmdline as user friendly as possible. 
It's not a top priority right now so I'll leave that in and we can clean up after all else is done.

**Comment by wacban** [comment]
https://nayduck.near.org/#/run/3273
fingers crossed

**Pull request merged**